File: src/main/java/planetGenerator/controllers/HeightGenViewController.java
package planetGenerator.controllers;

import javafx.beans.value.ChangeListener;
import javafx.beans.value.ObservableValue;
import javafx.fxml.FXML;
import javafx.scene.control.Label;
import javafx.scene.control.Slider;
import planetGenerator.generators.arguments.HeightArguments;
import planetGenerator.model.PlanetType;

public class HeightGenViewController {
	
	@FXML private Slider octavesSlider;
	@FXML private Slider roughnessSlider;
	@FXML private Slider scaleSlider;
	
	@FXML private Label octaves;
	@FXML private Label roughness;
	@FXML private Label scale;
	
	private boolean resetting;
	
	
	private MainMenuController mainMenuController;
	
	public void initialize(){
		initializeOctavesSlider();
		initializeRoughnessSlider();
		initializeScaleSlider();
	}
	
	private void initializeOctavesSlider() {
		
		octavesSlider.valueProperty().addListener(new ChangeListener<Number>() {
			@Override
			public void changed(ObservableValue<? extends Number> observable, Number oldValue, Number newValue) {
				if (!octavesSlider.valueChangingProperty().getValue()){
					getArgs().setOctaves(newValue.intValue());
					octaves.setText("" + newValue.intValue());
					reGenerate();
				}
			}
		});
	}
	
	private void initializeRoughnessSlider() {
		
		roughnessSlider.valueProperty().addListener(new ChangeListener<Number>() {
			@Override
			public void changed(ObservableValue<? extends Number> observable, Number oldValue, Number newValue) {
				if (!roughnessSlider.valueChangingProperty().getValue()){
					getArgs().setRoughness(newValue.floatValue());
					roughness.setText("" + newValue.floatValue());
					reGenerate();
				}
			}
		});
	}
	
	private void initializeScaleSlider() {
		
		scaleSlider.valueProperty().addListener(new ChangeListener<Number>() {
			@Override
			public void changed(ObservableValue<? extends Number> observable, Number oldValue, Number newValue) {
				if (!scaleSlider.valueChangingProperty().getValue()){
					getArgs().setScale(newValue.floatValue());
					scale.setText("" + newValue.floatValue());
					reGenerate();
				}
			}
		});
	}
	
	public void setMainController(MainMenuController mainMenuController) {
		this.mainMenuController = mainMenuController;
	}
	
	@FXML
	public void reset(){
		resetValues();
		reGenerate();
	}
	
	public void resetValues(){
		resetting = true;
		
		HeightArguments args = getInitialArgs();
		
		int initialOctaves = args.getOctaves();
		octavesSlider.setValue(initialOctaves);
		octaves.setText("" + initialOctaves);
		
		float initialRoughness = args.getRoughness();
		roughnessSlider.setValue(initialRoughness);
		roughness.setText("" + initialRoughness);
		
		float initialScale = args.getScale();
		scaleSlider.setValue(initialScale);
		scale.setText("" + initialScale);
		
		resetting = false;
	}
	
	private void reGenerate(){
		if (mainMenuController != null && !resetting){
			mainMenuController.createNewGenerator();
		}
	}
	
	private HeightArguments getArgs(){
		return mainMenuController.getGenerationArguments().getHeightArguments();
	}
	
	private HeightArguments getInitialArgs(){
		PlanetType type = PlanetType.EARTHLIKE;
		if (mainMenuController != null){
			type = mainMenuController.getCurrentPlanetType();
		}
		return type.getDefaultArguments().getHeightArguments();
	}
}

File: src/main/java/rak/planetGenerator/controllers/HeightGenViewController.kt
package rak.planetGenerator.controllers

import javafx.fxml.FXML
import javafx.scene.control.Label
import javafx.scene.control.Slider
import rak.planetGenerator.generators.arguments.HeightArguments
import rak.planetGenerator.model.PlanetType

class HeightGenViewController {

    @FXML private var octavesSlider: Slider? = null
    @FXML private var roughnessSlider: Slider? = null
    @FXML private var scaleSlider: Slider? = null

    @FXML private var octaves: Label? = null

    @FXML private var roughness: Label? = null
    @FXML private var scale: Label? = null

    private var resetting: Boolean = false


    private var mainMenuController: MainMenuController? = null

    private val args: HeightArguments
        get() = mainMenuController!!.generationArguments.heightArguments

    private val initialArgs: HeightArguments
        get() {
            var type = PlanetType.EARTHLIKE
            if (mainMenuController != null) {
                type = mainMenuController!!.currentPlanetType
            }
            return type.defaultArguments.heightArguments
        }

    fun initialize() {
        initializeOctavesSlider()
        initializeRoughnessSlider()
        initializeScaleSlider()
    }

    private fun initializeOctavesSlider() {

        octavesSlider!!.valueProperty().addListener { _, _, newValue ->
            if (octavesSlider?.valueChangingProperty()?.value != true) {
                args.octaves = newValue.toInt()
                octaves!!.text = "" + newValue.toInt()
                reGenerate()
            }
        }
    }

    private fun initializeRoughnessSlider() {

        roughnessSlider!!.valueProperty().addListener { _, _, newValue ->
            if (roughnessSlider?.valueChangingProperty()?.value != true) {
                args.roughness = newValue.toFloat()
                roughness!!.text = "" + newValue.toFloat()
                reGenerate()
            }
        }
    }

    private fun initializeScaleSlider() {

        scaleSlider!!.valueProperty().addListener { _, _, newValue ->
            if (scaleSlider?.valueChangingProperty()?.value != true) {
                args.scale = newValue.toFloat()
                scale!!.text = "" + newValue.toFloat()
                reGenerate()
            }
        }
    }

    fun setMainController(mainMenuController: MainMenuController) {
        this.mainMenuController = mainMenuController
    }

    @FXML
    fun reset() {
        resetValues()
        reGenerate()
    }

    fun resetValues() {
        resetting = true

        val args = initialArgs

        val initialOctaves = args.octaves
        octavesSlider!!.value = initialOctaves.toDouble()
        octaves!!.text = "" + initialOctaves

        val initialRoughness = args.roughness
        roughnessSlider!!.value = initialRoughness.toDouble()
        roughness!!.text = "" + initialRoughness

        val initialScale = args.scale
        scaleSlider!!.value = initialScale.toDouble()
        scale!!.text = "" + initialScale

        resetting = false
    }

    private fun reGenerate() {
        if (mainMenuController != null && !resetting) {
            mainMenuController!!.createNewGenerator()
        }
    }
}
